### PR TITLE
miner test suite refactor

### PR DESCRIPTION
**Pull request by andymck**
Refactor of common tests for code reuse and general flow improvements.

The following further items will be looked into ( if deemed to hold sufficient value )

- add specs to util functions

- further refactor to poc suite, use per init_per_suite, init_per_testcase and better following of existing conventions

- election test flakiness, fails sporadically 

- Implement marcs suggestion and move retries from wait until into assert so as to get better error reporting upon failure

- Check whether tests actually need the shuffle function

### CODE REVIEW

File: test/miner_poc_SUITE.erl [hunk 1]
```diff
@@ -3,6 +3,7 @@
 -include_lib("common_test/include/ct.hrl").
```
Comment (evanmcc) [inline test/miner_poc_SUITE.erl]: sorry to have wasted your time here, but the poc_SUITE has undergone major, incompatible revisions during the pocv4 PR.  can we just revert this whole thing to master for a clean merge both ways?  we can look at refactoring it later, it's done a lot of parallel work to this.

File: test/miner_blockchain_SUITE.erl [hunk 2]
```diff
@@ -93,270 +89,110 @@ init_per_testcase(TestCase, Config0) ->
...
-                                         P when is_pid(P) ->
-                                             true;
-                                         Other ->
-                                             ct:pal("Other ~p~n", [Other]),
-                                             false
-                                     end
-                             end, Miners)
-           end, 90, timer:seconds(1)),
-
+    ok = miner_ct_utils:wait_for_module(Miners, miner_consensus_mgr),
```
[the remark this answers is not in the document]
Comment (andymck) [reply test/miner_blockchain_SUITE.erl]: the function is a general purpose function which enables any registered process to be specified and await for that to return a PID.  Currently there are two case of it in use, this and another which waits for 'blockchain_worker'.  

Maybe it works better named as 'wait_for_process' ?

File: test/miner_blockchain_SUITE.erl [hunk 3]
```diff
@@ -688,18 +371,7 @@ group_change_test(Config) ->
...
-                                                              C1 = ct_rpc:call(Miner, blockchain_worker, blockchain, [], 500),
-                                                              L1 = ct_rpc:call(Miner, blockchain, ledger, [C1], 500),
-                                                              {ok, Sz} = ct_rpc:call(Miner, blockchain, config, [num_consensus_members, L1], 500),
-                                                              ct:pal("size ~p", [Sz]),
-                                                              true == ct_rpc:call(Miner, miner_consensus_mgr, in_consensus, [])
-                                                      end, Miners),
-                                           ct:pal("group size: ~p", [length(CGroup)]),
-                                           7 == length(CGroup)
-                                   end, 60, timer:seconds(1)),
+    ok = miner_ct_utils:wait_for_in_consensus(Miners, 7),
```
[the remark this answers is not in the document]
Comment (andymck) [reply test/miner_blockchain_SUITE.erl]: the implementation of this function actually already uses the new macro.  I got tired of looking at so many of those all over the code i moved as many as i thought sensible to the utils module, makes the main tests more readable.  

But this is the only usage of this function so happy to remove and use the macro directly if you think preferable

File: test/miner_ct_utils.erl [hunk 4]
```diff
@@ -14,16 +18,395 @@
...
-         random_n/2,
-         init_per_testcase/2,
-         end_per_testcase/2,
          get_balance/2,
          make_vars/1, make_vars/2, make_vars/3,
          tmp_dir/0, tmp_dir/1, nonl/1,
          generate_keys/1,
-         new_random_key/1
+         new_random_key/1,
+         stop_miners/1,
```
[the remark this answers is not in the document]
Comment (andymck) [reply test/miner_ct_utils.erl]: will do

File: test/miner_ct_utils.erl [hunk 5]
```diff
@@ -14,16 +18,395 @@
...
+            ct:pal("~p had Epoch ~p", [M, Other]),
+            timer:sleep(500),
+            consensus_members(Epoch, Tail)
+    catch C:E ->
+            ct:pal("~p threw error ~p:~p", [M, C, E]),
+            timer:sleep(500),
+            consensus_members(Epoch, Tail)
+    end.
+
+in_non_consensus_miners(Miners)->
```
[the remark this answers is not in the document]
Comment (andymck) [reply test/miner_ct_utils.erl]: agree

File: test/miner_poc_SUITE.erl (hunk 1 above)
Comment (andymck) [reply test/miner_poc_SUITE.erl]: will do

File: test/miner_ct_utils.erl [hunk 6]
```diff
@@ -14,16 +18,395 @@
...
+                             catch _:_ ->
+                                 false
+                             end
+                         end, Miners)
+                 end,
+        Result == true, 60, timer:seconds(5)),
+    ok.
+
+%% height can be that specified or greater
+wait_until_height(Miners, Height) ->
```
[the remark this answers is not in the document]
Comment (andymck) [reply test/miner_ct_utils.erl]: understood, will see if i can roll into a single function

File: test/miner_blockchain_SUITE.erl (hunk 2 above)
[the remark this answers is not in the document]
Comment (andymck) [reply test/miner_blockchain_SUITE.erl]: @evanmcc  renamed to wait_for_registration and moved to init_per_testcase as suggested.  The function is being called in two diff tests, each specifying a different process.  Is there a single process which can be checked which will be applicable to all miners irrespective of their mode ?

File: test/miner_blockchain_SUITE.erl [hunk 7]
```diff
@@ -766,17 +409,9 @@ master_key_test(Config) ->
...
-                   lists:all(
-                     fun(Miner) ->
-                             C = ct_rpc:call(Miner, blockchain_worker, blockchain, []),
-                             Ledger = ct_rpc:call(Miner, blockchain, ledger, [C]),
-                             {ok, totes_goats_garb} == ct_rpc:call(Miner, blockchain, config, [garbage_value, Ledger])
-                     end, shuffle(Miners))
-           end, 40, timer:seconds(1)),
-
+    
+    ok = miner_ct_utils:wait_for_txn_key_update(Miners, garbage_value, totes_goats_garb),
```
[the remark this answers is not in the document]
Comment (andymck) [reply test/miner_blockchain_SUITE.erl]: done

File: test/miner_ct_utils.erl (hunk 4 above)
[the remark this answers is not in the document]
Comment (andymck) [reply test/miner_ct_utils.erl]: done

File: test/miner_ct_utils.erl (hunk 5 above)
[the remark this answers is not in the document]
Comment (andymck) [reply test/miner_ct_utils.erl]: Renamed to miners_by_consensus_state, think its better but lemme know if you think it needs something else

File: test/miner_ct_utils.erl [hunk 8]
```diff
@@ -14,16 +18,395 @@
...
+            consensus_members(Epoch, Tail)
+    end.
+
+in_non_consensus_miners(Miners)->
+    lists:partition(
+            fun(Miner) ->
+                  true == ct_rpc:call(Miner, miner_consensus_mgr, in_consensus, [])
+            end, Miners).
+
+consensus_miners(Miners)->
```
[the remark this answers is not in the document]
Comment (andymck) [reply test/miner_ct_utils.erl]: renamed this to in_consensus_miners

File: test/miner_ct_utils.erl [hunk 9]
```diff
@@ -14,16 +18,395 @@
...
+    lists:filtermap(
+            fun(Miner) ->
+                true == ct_rpc:call(Miner, miner_consensus_mgr, in_consensus, [])
+            end, Miners).
+
+non_consensus_miners(Miners)->
```
[the remark this answers is not in the document]
Comment (andymck) [reply test/miner_ct_utils.erl]: Kept this as is, as it fits with the above renamed functions.

File: test/miner_ct_utils.erl [hunk 10]
```diff
@@ -14,16 +18,395 @@
...
+    lists:filtermap(
+            fun(Miner) ->
+                false == ct_rpc:call(Miner, miner_consensus_mgr, in_consensus, [])
+            end, Miners).
+
+election_check([], _Miners, Owner) ->
+    Owner ! seen_all;
+election_check(NotSeen0, Miners, Owner) ->
+    timer:sleep(500),
+    ConsensusMiners = miner_ct_utils:consensus_miners(Miners),
```
[the remark this answers is not in the document]
Comment (andymck) [reply test/miner_ct_utils.erl]: @evanmcc went over this one a few times now and I dont think I am gettting what you mean here.  Can you elaborate.

File: test/miner_poc_SUITE.erl (hunk 1 above)
Comment (andymck) [reply test/miner_poc_SUITE.erl]: done

File: test/miner_ct_utils.erl (hunk 6 above)
[the remark this answers is not in the document]
Comment (andymck) [reply test/miner_ct_utils.erl]: done

File: test/miner_blockchain_SUITE.erl (hunk 3 above)
[the remark this answers is not in the document]
Comment (andymck) [reply test/miner_blockchain_SUITE.erl]: As this is already utilising the new macro, i kept as is.

File: test/miner_ct_utils.erl (hunk 10 above)
[the remark this answers is not in the document]
Comment (andymck) [reply test/miner_ct_utils.erl]: Done

**Pull request merged**